docs/source/developer/concepts.rst

```diff
@@ -0,0 +1,307 @@
+.. Copyright (c) 2016, Johan Mabille, Sylvain Corlay and Wolf Vollprecht
+
+   Distributed under the terms of the BSD 3-Clause License.
+
+   The full license is in the file LICENSE, distributed with this software.
+
+.. _concepts-label:
+
+Concepts
+========
+
+`xtensor`'s core is built upon key concepts captured in interfaces that are put together in
+derived classes through CRTP and multiple inheritance. Interfaces and classes that model
+expressions implement value semantic. CRTP and value semantic achieve static polymorphism and
+avoids performance overhead of virtual methods and dynamic dispatching.
+
+xexpression
+~~~~~~~~~~~
+
+``xexpression`` is the base class for all expression classes. It is a CRTP base whose template
+parameter must be the most derived class in the hierarchy. For instance, if ``A`` inherits
+from ``B`` which in turn inherits from ``xexpression``, then ``B`` should be a template
+class whose template parameter is ``A`` and should forward this parameter to ``xexpression``:
+
+.. code::
+
+    #include "xtensor/xexpression.hpp"
+
+    template <class T>
+    class B : public xexpression<T>
+    {
+        // ...
+    };
+
+    class A : public B<A>
+    {
+        // ...
+    };
+
+``xexpression`` only provides three overloads of a same function, that cast an ``xexpression``
+object to the most inheriting type, depending on the nature of the object (lvalue,
+const lvalue or rvalue):
+
+.. code::
+
+    derived_type& derived_cast() & noexcept;
+    const derived_type& derived_cast() & noexcept;
+    derived_type derived_cast() && noexcept;
+
+.. _xiterable-concept-label:
+
+xiterable
+~~~~~~~~~
+
+The iterable concept is modeled by two classes, ``xconst_iterable`` and ``xiterable``, defined
+in ``xtensor/xiterable.hpp``. ``xconst_iterable`` provides types and methods for iterating on
+constant expressions, similar to the ones provided by the STL containers. Unlike the STL, the
+methods of ``xconst_iterable`` and ``xiterable`` are templated by a layout parameter that allows
+you to iterate over a N-dimensional expression in row-major or column-major.
```

Comment:
in row-major or column-major order.

Row-major iteration means that the right-most indices vary faster.
Column major iteration means that the left-most indices vary faster.

Or something like...

Row-major layout means that elements that only differ by their last index are contiguous in memory.
Column-major layout means that elements that only differ by their first index are contiguous in memory.